ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs

```diff
@@ -264,47 +271,268 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
+-- In case no successors were found, the original 'ChainDiff' is returned as a
+-- singleton.
 --
--- PRECONDITION: @B ∈ V@, i.e. @B@ is part of the VolatileDB.
+-- In case successors /were/ found, the original 'ChainDiff' is /not/
+-- included, only its extensions.
 --
--- 'True' <=> for all transitive predecessors @B'@ of @B@ we have @B' ∈ V@ or
--- @B' = I@.
-isReachable :: forall blk. (HasHeader blk, HasCallStack)
-            => (HeaderHash blk -> Maybe (WithOrigin (HeaderHash blk)))
-            -> Point blk                                        -- ^ @I@
-            -> RealPoint blk                                    -- ^ @B@
-            -> Maybe (NonEmpty (HeaderHash blk))
-isReachable predecessor i b =
-    case computePath predecessor from to of
-      -- Bounds are not on the same fork
-      Nothing   -> Nothing
-      Just path -> case path of
-        CompletelyInVolDB hashes -> case NE.nonEmpty hashes of
-          Just hashes' -> Just hashes'
-          Nothing      ->
-            error "impossible: empty list of hashes with an inclusive bound"
-        PartiallyInVolDB  {}     -> Nothing
-        NotInVolDB        _      ->
-          error "impossible: block just added to VolatileDB missing"
+-- Only the longest possible extensions are returned, no intermediary prefixes
+-- of extensions.
+extendWithSuccessors ::
+     forall blk. HasHeader blk
+  => (WithOrigin (HeaderHash blk) -> Set (HeaderHash blk))
+  -> LookupBlockInfo blk
+  -> ChainDiff (HeaderFields blk)
+  -> NonEmpty (ChainDiff (HeaderFields blk))
+extendWithSuccessors succsOf lookupBlockInfo diff =
+    case NE.nonEmpty extensions of
+      Nothing          -> diff NE.:| []
+      Just extensions' -> extensions'
   where
-    from = StreamFromExclusive i
-    to   = StreamToInclusive   b
+    extensions =
+        [ foldl' Diff.append diff (lookupHeaderFields <$> candHashes)
+        | candHashes <- candidates succsOf (Diff.getTip diff)
+        ]
+
+    lookupHeaderFields :: HeaderHash blk -> HeaderFields blk
+    lookupHeaderFields =
+          headerFieldsFromBlockInfo
+          -- The successor mapping is populated with the blocks in the
+          -- VolatileDB, so looking up the block info of a successor /must/
+          -- succeed.
+        . fromMaybe (error "successor must in the VolatileDB")
+        . lookupBlockInfo
+
+{-------------------------------------------------------------------------------
+  Paths through the VolatileDB
+-------------------------------------------------------------------------------}
+
+-- | Return the block info for the block with the given hash. Return 'Nothing'
+-- when not in the VolatileDB.
+type LookupBlockInfo blk =
+  HeaderHash blk -> Maybe (VolDB.BlockInfo (HeaderHash blk))
+
+headerFieldsFromBlockInfo ::
+     VolDB.BlockInfo (HeaderHash blk)
+  -> HeaderFields blk
+headerFieldsFromBlockInfo VolDB.BlockInfo { bslot, bbid, bbno } =
+    HeaderFields {
```

Comment:
Try making the fields strict. If that works, use it in `VolDB.BlockInfo` and `Anchor`.

Reply:
Let's do that in a separate PR.